### ISSUE TITLE
Store old pages data and show them in place of the white loading

### ISSUE THREAD

**Issue by akramsarkis** [Geeks-Solutions/nuxt-sections#100]
- [ ] when you go from Page A to Page B, instead of showing a white page, you keep on displaying Page A until you received the response for Page B
- This is an option to avoid displaying a white screen in between the pages, another option could be to render the first fold and then render the rest of the page after. This option though would only be valid if a significant amount of time is spent rendering the page (rendering: building the HTML after we received the API response from Sections Backend)
- [ ] Automated test to be done

### Acceptance testing

- [ ] When I switch from a page A to a page B, till the response for page B is received, you keep on displaying page A
- [ ] Automated test to be done

**Comment by jfayad** [comment]
Are you sure about the spec ?

Nothing needs to be stored... it's just that when you go from Page A to Page B
Instead of showing a white page, you keep on displaying Page A until you received the response for Page B

This is an option to avoid displaying a white screen in between the pages, another option could be to render the first fold and then render the rest of the page after. This option though would only be valid if a significant amount of time is spent rendering the page (rendering: building the HTML after we received the API response from Sections Backend)

**Comment by akramsarkis** [comment]
@mahfouzhsein I updated the description so if you can make sure of the estimate

**Comment by mahfouzhseineweev** [comment]
### Report:

There are multiple options we can use to achieve these results:

* Option 1) Use localStorage to set data when received from the api and to get them when navigating from a page to another

Test during the Timebox but the results were not good.
As localStorage doesn't work on server side it can not be initialized inside the data() property of the section component.
So I tried setting it inside the mounted hook which worked, yet there is delay before the new page show and this delay shows a white page and not the content of the store page

* Option 2) Use $cookies with a 1h expiry maybe, which works in server side and can be directly passed inside data for the views

Still did haven't tested this one

* Option 3) Install Vuex store and use it to persist data when navigating through the pages.

Vuex is not installed currently on the library. This option could take a bit more time to be implemented

---

I will need 1 final Timebox to test Option 2 and make sure if it is valid and solid one that we can use

**Comment by jfayad** [comment]
I don't understand what are you doing here ?

1. Why is SSR relevant here ? We are specifically talking about a browsing experience after the user opened a page already 

2. Why are looking for a solution to store data ? I already left a comment stating this is not a good approach in my opinion and your timebox is just confirming that...

**Comment by jfayad** [comment]
> @jfayad These are the solutions I found, they are all pointing into that direction (storing data)

@mahfouzhseineweev what do you mean by solutions you found ? You'll find what you search for... the goal is to keep the display of the current page while you load the new page to open so what exactly do you need to store ?

All is required now is SIMPLY to not show a blank page before you finish loading the new page, this is a trivial issue that is solved out of box in nuxt apps but it seems the library's management of page navigation is "breaking" this management and all you need to do is restore that.

You also ignored my first question about SSR ?

**Comment by mahfouzhseineweev** [comment]
> You also ignored my first question about SSR ?

But this is no longer relevant as the Timebox I did was not valid.

> All is required now is SIMPLY to not show a blank page before you finish loading the new page, this is a trivial issue that is solved out of box in nuxt apps but it seems the library's management of page navigation is "breaking" this management and all you need to do is restore that.

@jfayad  From inside page A, awaiting for an API response call to be completed on page B before performing the navigation to it.
You sure this is solved out of box in nuxt apps ?

**Comment by jfayad** [comment]
> > You also ignored my first question about SSR ?
> 
> But this is no longer relevant as the Timebox I did was not valid.
> 
> > All is required now is SIMPLY to not show a blank page before you finish loading the new page, this is a trivial issue that is solved out of box in nuxt apps but it seems the library's management of page navigation is "breaking" this management and all you need to do is restore that.
> 
> @jfayad From inside page A, awaiting for an API response call to be completed on page B before performing the navigation to it. You sure this is solved out of box in nuxt apps ?

Are you aware of how nuxt performs a link click out of the box ?

**Comment by mahfouzhseineweev** [comment]
@jfayad I think you are referring to this https://v2.nuxt.com/docs/features/nuxt-components#prefetchlinks
Which is what I should be aiming to check and fix in my Timebox

**Comment by jfayad** [comment]
> @jfayad I think you are referring to this https://v2.nuxt.com/docs/features/nuxt-components#prefetchlinks Which is what I should be aiming to check and fix in my Timebox

I'm asking you a general question about how Nuxt handles link click.
My high level understanding is that the loading of the new page happens while still on the current page, which is the opposite of what you are suggesting (to be on the new page and display the old page there...)

**Comment by mahfouzhseineweev** [comment]
@jfayad In order to validate the theory that nuxt section library is breaking the navigation management, I went to sections BO and checked the navigation there and it is the same behavior, the page is opened with empty data and then the API request is called with a loading indicator.

Now I know that using a middleware in pageB will pause the navigation until the response of API call in the middleware is returned and more of this using asyncData in pageB will also pause the navigation to it until the response of the API is returned which is what we want to achieve in the library.

The library is using the fetch hook and asynchronously call the API to get a page which should work similarly to the asyncData.

In the Timebox that I will do, this is what I am planning to figure out.

=> Why using an async fetch hook is not working like the asyncData and so breaking the nuxt navigation

**Comment by jfayad** [comment]
> @jfayad In order to validate the theory that nuxt section library is breaking the navigation management, I went to sections BO and checked the navigation there and it is the same behavior, the page is opened with empty data and then the API request is called with a loading indicator.
> 
> Now I know that using a middleware in pageB will pause the navigation until the response of API call in the middleware is returned and more of this using asyncData in pageB will also pause the navigation to it until the response of the API is returned which is what we want to achieve in the library.
> 
> The library is using the fetch hook and asynchronously call the API to get a page which should work similarly to the asyncData.
> 
> In the Timebox that I will do, this is what I am planning to figure out.
> 
> => Why using an async fetch hook is not working like the asyncData and so breaking the nuxt navigation

LGTM

**Comment by mahfouzhseineweev** [comment]
### Reprot:

@jfayad  I gave this a 30mn validate the theory.

The difference in behavior we are experiencing is because of how async fetch() and asyncData work in Nuxt.js:

1) asyncData:
This method is executed during server-side rendering (SSR) or before the route navigation is completed when running on the client. Nuxt waits for the asyncData method to resolve before rendering the page or completing the navigation.

2) fetch:
The fetch method, is designed to fetch data when the component is loaded, but it does not block navigation. Nuxt navigates to the page immediately and executes fetch in the background.


In the library we are using the fetch hook because asyncData can only be used in page level while sections code is a component level. So what is happening is not a bug or broken feature but it is how fetch works in nuxt

As an alternative to the fetch we can use the beforeRouteEnter navigation guard which can be used to delay navigation until data is fetched. This guard blocks navigation by design but is not ideal for SSR since it runs only on the client side.

Note that beforeRouteEnter is a bit tricky because it doesn't have direct access to this

```
beforeRouteEnter(to, from, next) {
    // Fetch the data before navigating
    const axios = require('axios');
    axios
      .get('https://api.example.com/resource')
      .then((response) => {
        next((vm) => {
          vm.dataFetched = response.data;
        });
      })
      .catch((error) => {
        console.error('Failed to fetch data:', error);
        next(); // Allow navigation even if the request fails
      });
  },
  ```

So I think we can use a mix of fetch ssr (the one we have) and beforeRouteEnter for csr

**Comment by jfayad** [comment]
> So I think we can use a mix of fetch ssr (the one we have) and beforeRouteEnter for csr

You keep on referring to SSR, but I really don't understand your concern here... is there something I'm overlooking ?

When the page loads in SSR it means there's no page currently open on the browser, so how is SSR relevant for this issue ?

**Comment by mahfouzhseineweev** [comment]
@jfayad The only why I am referring to SSR is that I am currently using the fetch() hook for both SSR and CSR and so any change or alternative I am going to use should be considering SSR too to make sure this functionality is not broken with this update

**Comment by jfayad** [comment]
Got you, indeed, LGTM

**Comment by mahfouzhseineweev** [comment]
### Report:

Things did not go as expected at all.

* The beforeRouteEnter did not work in the component level as it is designed for page level

* It seems there is no other way in nuxt to block the navigation other than using the asyncData hook

* If we considered using asyncData for csr inside the dynamic page `_.vue` and pass the data to the Sections Main component to process it then it will only work projects that are using dynamic page management and so declaring a page from the host project will no longer work.

* Even if we decided to go with the asyncData hook, it wont be an easy and clean solution but it solve the issue


@jfayad I feel I am reaching a dead end here, can you help me ?

**Comment by jfayad** [comment]
> If we considered using asyncData for csr inside the dynamic page _.vue and pass the data to the Sections Main component to process it then it will only work projects that are using dynamic page management and so declaring a page from the host project will no longer work.

Can you explain why exactly ?

It seems to me like the host project could just use the asyncData on the new page it would be willing to use, couldn't it ?

**Comment by jfayad** [comment]
> @jfayad Indeed, the host project can use the asyncData to run the render page API and pass the response as a prop to the Sections component.
> 
> Then asyncData will be in charge of getting the page data on client side and passing the response to the sections component as a prop.
> 
> Sections component will read the response and initialize the sections.
> 
> The SSR management will remain in the Sections component fetch hook.

the only issue I can see here is that the host project would then have to be issuing the API request to the backend...

Could we provide a helper function that the host project dev. would just need to stitch inside the asyncData implementation ?

**Comment by mahfouzhseineweev** [comment]
> Could we provide a helper function that the host project dev. would just need to stitch inside the asyncData implementation ?

@jfayad  Yes, this should be possible, the function will return the page data and the host project in the asyncData will feed the sections prop by the value returned

**Comment by jfayad** [comment]
if I wrap up, we can solve the blank page issue by using asyncData and that would be totally transparent when using the _pageName.vue dynamic page as the API call to render a page will move from the sections component to the page

that would require an extra management for the host project developer who would want to use Sections inside a separate page component by issuing the request inside the asyncData...

Why don't we keep the current management of issuing request so the component could be used as is inside another page, and have a prop. that would move the page request to the asyncData, which would be used by the dynamic page ?

that way a host project dev. could choose to have an out of the box approach (the one we have now) vs a more involving approach relying on the asyncData.

any thoughts @mahfouzhseineweev ?

**Comment by mahfouzhseineweev** [comment]
@jfayad 

I think no additional prop is needed for asyncData.

Basically I can introduce a sections array prop in the Sections component that will be used to get the sections from the asyncData used in the dynamic page or the host project page.

Then inside the fetch hook CSR management, I read the value of this sections array prop and if it is empty (Host project did not implement the asyncData hook), the current way we have will be still in place (request sent in CSR to get the page and the white page will still be showing).

If if the sections array prop has a value (coming from the dynamic page asyncData), The fetch hook will just initialize the sections from this prop

**Comment by jfayad** [comment]
> Basically I can introduce a sections array prop in the Sections component that will be used to get the sections from the asyncData used in the dynamic page or the host project page.

Would that be the responsibility of the host project to get the data and put them in the right prop or can a new helper function store the data from the api call to pass it to the fetch hook down the line ?

> If if the sections array prop has a value (coming from the dynamic page asyncData), The fetch hook will just initialize the sections from this prop

Ok, that also means that in SSR there will be no impact as data will still be loaded inside the fetch hook ?

**Comment by mahfouzhseineweev** [comment]
@jfayad 

> Would that be the responsibility of the host project to get the data and put them in the right prop or can a new helper function store the data from the api call to pass it to the fetch hook down the line ?

Yes the host project will first call the helper function (Options and Post requests to get the page) that will return the sections data.

The host project will use the returned sections and feed then to the <Sections /> component using the new sections prop

> Ok, that also means that in SSR there will be no impact as data will still be loaded inside the fetch hook ?

Correct

**Comment by jfayad** [comment]
> Yes the host project will first call the helper function (Options and Post requests to get the page) that will return the sections data.

The host project will use the returned sections and feed then to the component using the new sections prop

@mahfouzhseineweev can you write a small code sample in here to see concretely how involving that would be for the host project dev. ?

**Comment by mahfouzhseineweev** [comment]
@jfayad I marked the lines to be added in between lines `--------`

```vue
<template>
  <Sections
    :admin="admin"
    :page-name="pageName"
    :lang="lang"
    :variations="[]"
    --------
    :sections-data="sectionsData"
    --------
  />
</template>

<script>
--------
import {helperFunctionToGetReponseOfRenderSectionsPage} from "@geeks.solutions/nuxt-sections/lib/src/utils";
--------

export default {
  name: 'IndexPage',
  data() {
    return {
      pageName: "Home",
      --------
      sectionsData: []
      --------
    };
  },
  computed: {
    lang() {
      return this.$i18n.locale.toString()
    },
    admin() {
      return !!this.$cookies.get("sections-auth-token")
    }
  },
  --------
  async asyncData({ params, $axios }) {
    return {
      sectionsData: helperFunctionToGetReponseOfRenderSectionsPage().data.sections || [], // Return the fetched data or a 
        default value
    }
  }
  --------
}
</script>
```

**Comment by jfayad** [comment]
@mahfouzhseineweev LGTM

**Comment by mahfouzhseineweev** [comment]
Estimation range: 2.5h to 3.5h

**Comment by mahfouzhseineweev** [comment]
I have made a great progress on the issue and sent a PR on the library for it, but I am still not done with the issue (spent 1h12mn)

**Comment by jfayad** [comment]
what exactly is left to be done here ?

**Comment by mahfouzhseineweev** [comment]
Nothing else to be done for this issue. It can be closed.
